### PR TITLE
Create Target Activity

### PR DESCRIPTION
**Pull request by marcossalto**
Create Target Activity, add map fragment and show device location.

Screenshot_20210224-131124_Target MVD

### CODE REVIEW

File: app/src/main/java/com/marcossalto/targetmvd/network/managers/ILocationManager.kt [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+import android.content.Context
+import android.location.Location
+
+interface ILocationManager {
+    fun saveLocation(lat: Double, lng: Double)
+    fun isStateSuccess(): Boolean
+    fun getLatitude(): Double
+    fun getLongitude(): Double
+    fun getLocation(context: Context, successActionCallback: (location: Location) -> Unit)
+}
```
Comment (amaury901130) [inline app/src/main/java/com/marcossalto/targetmvd/network/managers/ILocationManager.kt]: add an empty line

File: app/build.gradle [hunk 2]
```diff
@@ -112,10 +121,16 @@ dependencies {
...
     // Facebook Login
     implementation 'com.facebook.android:facebook-login:8.1.0'
 
+    // Google Maps
+    implementation 'com.google.android.gms:play-services-maps:17.0.0'
+
+    // Location
+    implementation "com.google.android.gms:play-services-location:17.1.0"
+
 }
```
Comment (amaury901130) [inline app/build.gradle]: add an extra empty line

File: app/src/main/java/com/marcossalto/targetmvd/network/managers/LocationManager.kt [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import com.google.android.gms.location.LocationServices
+
+object LocationManager : ILocationManager {
+
+    private var fusedLocationProviderClient: FusedLocationProviderClient? = null
+
+    override fun saveLocation(lat: Double, lng: Double) {
+        UserLocation.latitude = lat
+        UserLocation.longitude = lng
+        UserLocation.locationState = LocationState.SUCCESS
```
Comment (amaury901130) [inline app/src/main/java/com/marcossalto/targetmvd/network/managers/LocationManager.kt]: Use the scope function `.apply ` to assign values over the same object.
```suggestion
        UserLocation.apply {
            latitude = lat
            longitude = lng
            locationState = LocationState.SUCCESS
        }
```
please check this post in the Rootstrap blog, about the scope functions
https://www.rootstrap.com/blog/scope-functions-in-kotlin/

File: app/src/main/java/com/marcossalto/targetmvd/ui/base/BaseFragment.kt [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        }
+
+        builder.setPositiveButton(getString(R.string.ok)) { dialog, _ ->
+            dialog.cancel()
+        }
+
+        val dialog: AlertDialog = builder.create()
+        dialog.show()
+    }
+}
```
Comment (amaury901130) [inline app/src/main/java/com/marcossalto/targetmvd/ui/base/BaseFragment.kt]: add empty line.

File: app/src/main/java/com/marcossalto/targetmvd/ui/base/BaseFragment.kt [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+        when (message) {
+            "" -> builder.setMessage(getString(R.string.generic_error))
+            null -> builder.setMessage(getString(R.string.generic_error))
+            else -> builder.setMessage(message)
+        }
+
+        builder.setPositiveButton(getString(R.string.ok)) { dialog, _ ->
+            dialog.cancel()
+        }
```
Comment (amaury901130) [inline app/src/main/java/com/marcossalto/targetmvd/ui/base/BaseFragment.kt]: try to use a scope function with the builder
I'd use something like this to assign the message.
```
val builderMessage = if(message.isNullOrBlank()) 
      getString(R.string.generic_error) 
  else 
      message
```

File: app/src/main/java/com/marcossalto/targetmvd/ui/target/TargetActivity.kt [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    private lateinit var viewModel: TargetActivityViewModel
+    private lateinit var binding: ActivityTargetBinding
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        binding = ActivityTargetBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+
+        Analytics.track(PageEvents.visit(VISIT_TARGET
+        ))
```
Comment (amaury901130) [inline app/src/main/java/com/marcossalto/targetmvd/ui/target/TargetActivity.kt]: check the parenthesis

File: app/src/main/java/com/marcossalto/targetmvd/ui/target/TargetActivity.kt [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+            )
+        }
+
+    }
+
+    private fun initMap() {
+        val fragment = MapFragment.getInstance()
+        supportFragmentManager.beginTransaction().replace(R.id.fragment_map, fragment).commit()
+    }
+}
```
Comment (amaury901130) [inline app/src/main/java/com/marcossalto/targetmvd/ui/target/TargetActivity.kt]: add empty line

File: app/src/main/java/com/marcossalto/targetmvd/ui/target/TargetActivityViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    fun getLocation(context: Context, successAction: (location: Location) -> Unit) {
+        locationManager.getLocation(context, successAction)
+    }
+}
+
+class TargetActivityViewModelFactory() : ViewModelProvider.Factory {
+    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
+        return TargetActivityViewModel(LocationManager)  as T
+    }
+}
```
Comment (amaury901130) [inline app/src/main/java/com/marcossalto/targetmvd/ui/target/TargetActivityViewModel.kt]: same

File: app/src/main/res/layout/activity_sign_in.xml [hunk 9]
```diff
@@ -125,25 +125,12 @@
...
-            app:layout_constraintEnd_toEndOf="parent"
-            app:layout_constraintStart_toStartOf="parent"
-            app:layout_constraintTop_toTopOf="parent" /-->
-
         <com.facebook.login.widget.LoginButton
             android:id="@+id/login_button"
             android:layout_width="wrap_content"
             android:layout_height="wrap_content"
             android:layout_gravity="center_horizontal"
             android:layout_marginTop="415dp"
```
Comment (amaury901130) [inline app/src/main/res/layout/activity_sign_in.xml]: refactor this, and use @dimen/

File: app/src/main/res/layout/fragment_map.xml [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+android:layout_height="match_parent"
+tools:context=".ui.target.MapFragment">
+
+<fragment
+    android:id="@+id/map"
+    android:name="com.google.android.gms.maps.SupportMapFragment"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    tools:context=".ui.target.TargetActivity" />
+</FrameLayout>
```
Comment (amaury901130) [inline app/src/main/res/layout/fragment_map.xml]: add an empty line.

File: app/src/main/res/values/colors.xml [hunk 11]
```diff
@@ -9,4 +9,6 @@
     <color name="white">#FFFFFFFF</color>
     <color name="red">#E02424</color>
     <color name="background">#F2F2F2</color>
+    <color name="yellow">#EFC638</color>
+    <color name="white_60">#99FFFFFF</color>
 </resources>
```
Comment (amaury901130) [inline app/src/main/res/values/colors.xml]: same.

File: app/src/main/res/values/strings.xml [hunk 12]
```diff
@@ -33,4 +33,6 @@
     <string name="select_your_gender">SELECT YOUR GENDER</string>
     <string name="facebook_app_id">876877256214170</string>
     <string name="fb_login_protocol_scheme">fb876877256214170</string>
+    <string name="create_new_target">CREATE NEW TARGET</string>
+    <string name="target_points">Target Points</string>
 </resources>
```
Comment (amaury901130) [inline app/src/main/res/values/strings.xml]: same

**Review by amaury901130 — CHANGES_REQUESTED** [review]
👍 good, just a couple of suggestions
for the next time, try to do a small PR.
also would be nice to have functions comments.

File: app/src/main/java/com/marcossalto/targetmvd/network/managers/LocationManager.kt (hunk 3 above)
Comment (marcossalto) [reply app/src/main/java/com/marcossalto/targetmvd/network/managers/LocationManager.kt]: Thanks for your suggestion. I am going to apply it

File: app/src/main/java/com/marcossalto/targetmvd/network/managers/ILocationManager.kt (hunk 1 above)
Comment (marcossalto) [reply app/src/main/java/com/marcossalto/targetmvd/network/managers/ILocationManager.kt]: Thanks for your suggestion

File: app/build.gradle (hunk 2 above)
Comment (marcossalto) [reply app/build.gradle]: Thanks for your suggestion

File: app/src/main/java/com/marcossalto/targetmvd/ui/base/BaseFragment.kt (hunk 5 above)
Comment (marcossalto) [reply app/src/main/java/com/marcossalto/targetmvd/ui/base/BaseFragment.kt]: Thanks for your suggestion

File: app/src/main/java/com/marcossalto/targetmvd/ui/base/BaseFragment.kt (hunk 4 above)
Comment (marcossalto) [reply app/src/main/java/com/marcossalto/targetmvd/ui/base/BaseFragment.kt]: Thanks for your suggestion

File: app/src/main/java/com/marcossalto/targetmvd/ui/target/TargetActivity.kt (hunk 6 above)
Comment (marcossalto) [reply app/src/main/java/com/marcossalto/targetmvd/ui/target/TargetActivity.kt]: Thanks for your suggestion

File: app/src/main/java/com/marcossalto/targetmvd/ui/target/TargetActivity.kt (hunk 7 above)
Comment (marcossalto) [reply app/src/main/java/com/marcossalto/targetmvd/ui/target/TargetActivity.kt]: Thanks for your suggestion

File: app/src/main/java/com/marcossalto/targetmvd/ui/target/TargetActivityViewModel.kt (hunk 8 above)
Comment (marcossalto) [reply app/src/main/java/com/marcossalto/targetmvd/ui/target/TargetActivityViewModel.kt]: Thanks for your suggestion

File: app/src/main/res/values/strings.xml (hunk 12 above)
Comment (marcossalto) [reply app/src/main/res/values/strings.xml]: Thanks for your suggestion

File: app/src/main/res/values/colors.xml (hunk 11 above)
Comment (marcossalto) [reply app/src/main/res/values/colors.xml]: Thanks for your suggestion

File: app/src/main/res/layout/fragment_map.xml (hunk 10 above)
Comment (marcossalto) [reply app/src/main/res/layout/fragment_map.xml]: Thanks for your suggestion

File: app/src/main/res/layout/activity_sign_in.xml (hunk 9 above)
Comment (marcossalto) [reply app/src/main/res/layout/activity_sign_in.xml]: Thanks for your suggestion. I am going to improve it

**Comment by marcossalto** [comment]
> 👍 good, just a couple of suggestions
> for the next time, try to do a small PR.
> also would be nice to have functions comments.

File: app/src/main/res/layout/activity_sign_up.xml [hunk 13]
```diff
@@ -206,81 +198,77 @@
...
             app:layout_constraintStart_toStartOf="parent"
-            app:layout_constraintTop_toTopOf="parent" />
+            app:layout_constraintTop_toBottomOf="@+id/confirm_password_edit_text" />
 
         <TextView
             android:id="@+id/gender_text_view_sign_up"
             android:layout_width="wrap_content"
             android:layout_height="wrap_content"
-            android:layout_marginTop="376dp"
+            android:layout_marginTop="18dp"
```
Comment (amaury901130) [inline app/src/main/res/layout/activity_sign_up.xml]: use `@dimen/`

File: app/src/main/res/layout/activity_sign_up.xml [hunk 14]
```diff
@@ -206,81 +198,77 @@
...
-            app:layout_constraintTop_toTopOf="parent" />
+            app:layout_constraintTop_toBottomOf="@+id/gender_spinner" />
 
         <Button
             android:id="@+id/sign_up_button_sign_up"
             style="@style/Widget.AppCompat.Button"
             android:layout_width="wrap_content"
             android:layout_height="wrap_content"
-            android:layout_marginTop="465dp"
+            android:layout_marginTop="19dp"
```
Comment (amaury901130) [inline app/src/main/res/layout/activity_sign_up.xml]: same

**Review by amaury901130 — APPROVED** [review]

File: app/src/main/res/layout/activity_sign_up.xml (hunk 13 above)
Comment (marcossalto) [reply app/src/main/res/layout/activity_sign_up.xml]: Thanks for your suggestion

File: app/src/main/res/layout/activity_sign_up.xml (hunk 14 above)
Comment (marcossalto) [reply app/src/main/res/layout/activity_sign_up.xml]: Thanks for your suggestion

**Review by agustinkoll-rootstrap — APPROVED** [review]

**Pull request merged**